File: leet_BST_108_sorted_array_to_btree___2pointer.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def make_tree(nums, L, R):
            if L > R:
                return None
            midpoint = (R - L) // 2 + L
            temp = TreeNode(nums[midpoint])
            temp.left = make_tree(nums, L, midpoint - 1)
            temp.right = make_tree(nums, midpoint + 1, R)
            return temp
        return make_tree(nums, 0, len(nums) - 1)

File: leet_BST_108_sorted_array_to_btree___2pointer.java

class Solution {
    public TreeNode sortedArrayToBST(int[] nums)
    {
        return make_tree(nums, 0, nums.length - 1);
    }

    public TreeNode make_tree(int[] n, int L, int R)
    {
        TreeNode    node;
        int         mid;

        if (L > R)
            return null;
        mid = ((R - L) >> 1) + L;
        node = new TreeNode(n[mid]);
        node.left = make_tree(n, L, mid - 1);
        node.right = make_tree(n, mid + 1, R);
        return node;
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
